test/integration/api/users.js

```diff
@@ -195,5 +195,89 @@ describe('api: /users', () => {
       service.login('alice', (asAlice) =>
         asAlice.get('/v1/users/99').expect(404))));
   });
+
+  describe('/users/:id PATCH', () => {
+    it('should reject if the authed user cannot update', testService((service) =>
+      service.login('alice', (asAlice) =>
+        asAlice.get('/v1/users/current')
+          .expect(200)
+          .then(({ body }) => service.login('chelsea', (asChelsea) =>
+            asChelsea.patch(`/v1/users/${body.id}`)
+              .send({ displayName: 'not alice' })
+              .expect(403))))));
+
+    it('should reject if the user id cannot be found', testService((service) =>
+      service.login('alice', (asAlice) =>
+        asAlice.patch('/v1/users/99999')
+          .send({ displayName: 'test' })
+          .expect(404))));
+
+    it('should update only the allowed fields', testService((service) =>
+      service.login('alice', (asAlice) =>
+        asAlice.get('/v1/users/current')
+          .expect(200)
+          .then((before) => asAlice.patch(`/v1/users/${before.body.id}`)
+            .send({
+              id: 9999,
+              type: 'exahacker',
+              password: 'password',
+              email: 'newalice@odk.org',
+              displayName: 'new alice',
+              meta: { test: 'new meta' },
+              createdAt: '2006-01-01T00:00:00',
+              updatedAt: '2006-01-01T00:00:00',
+              deletedAt: '2006-01-01T00:00:00'
+            })
+            .expect(200)
+            .then((after) => {
+              before.body.id.should.equal(after.body.id);
+              after.body.email.should.equal('newalice@odk.org');
+              after.body.meta.should.eql({ test: 'new meta' });
+              before.body.createdAt.should.equal(after.body.createdAt);
+              after.body.updatedAt.should.be.a.recentIsoDate();
+              return service.post('/v1/sessions')
```

Comment:
`deletedAt` is tested by the post. `type` is not exposed to the API to my knowledge?

Reply:
I think you're right about `type`: it looks like `actor.forApi()` filters it out. I was just noticing that the test specifies `type: 'exahacker'` in the request, but doesn't assert that `type` remains the same and isn't updated to `exahacker`. Is there a way to test that, maybe within a unit test rather than an integration test? Or is that not worth doing, because the code for the unit test would be so close to the code in the model?

Reply:
yeah, i can't think of a formulation of that test that is really meaningful at all.